### PR TITLE
Add fallback URL to handoff

### LINKED ISSUE #5566: Support Handoff fallback to website
From https://developer.apple.com/library/ios/documentation/UserExperience/Conceptual/Handoff/AdoptingHandoff/AdoptingHandoff.html#//apple_ref/doc/uid/TP40014338-CH2-SW19:

> When using a native app on the originating device, the user may want to continue the activity on another device that does not have a corresponding native app. If there is a web page that corresponds to the activity, it can still be handed off. For example, video library apps enable users to browse movies available for viewing, and mail apps enable users to read and compose email, and in many cases users can do the same activity though a web-page interface. In this case, the native app knows the URL for the web interface, possibly including syntax designating a particular video being browsed or message being read. **So, when the native app creates the NSUserActivity object, it sets the webpageURL property, and if the receiving device doesn't have an app that supports the user activity’s activityType, it can resume the activity in the default web-browser of the continuing platform.** :heart:

You can read more about the property at: https://developer.apple.com/library/ios/documentation/Foundation/Reference/NSUserActivity_Class/index.html#//apple_ref/occ/cl/NSUserActivity/

### PR DESCRIPTION
**Pull request by danhp**
Closes #5566

### CODE REVIEW

File: atom/browser/browser_mac.mm [hunk 1]
```diff
@@ -112,12 +112,16 @@
 void Browser::SetAppUserModelID(const base::string16& name) {
 }
 
-void Browser::SetUserActivity(
-    const std::string& type,
-    const base::DictionaryValue& user_info) {
+void Browser::SetUserActivity(const std::string& type,
+                              const base::DictionaryValue& user_info,
+                              mate::Arguments* args) {
+  std::string URL = "";
```
Comment (zcbenz) [inline atom/browser/browser_mac.mm]: Passing string directly may have [escaping problems](https://github.com/electron/electron/issues/2768), the safe way is to use `net::NSURLWithGURL` to convert the url to NSURL.

File: atom/browser/browser_mac.mm [hunk 2]
```diff
@@ -112,12 +112,16 @@
...
+  args->GetNext(&URL);
+
   [[AtomApplication sharedApplication]
       setCurrentActivity:base::SysUTF8ToNSString(type)
-            withUserInfo:DictionaryValueToNSDictionary(user_info)];
+            withUserInfo:DictionaryValueToNSDictionary(user_info)
+          withWebpageURL:[NSURL URLWithString:base::SysUTF8ToNSString(URL)]];
```
Comment (zcbenz) [inline atom/browser/browser_mac.mm]: We can probably pass `nil` here if the `url` is not specified.

File: atom/browser/browser_mac.mm [hunk 3]
```diff
@@ -112,12 +114,16 @@
...
+                              const base::DictionaryValue& user_info,
+                              mate::Arguments* args) {
+  std::string url_string;
+  args->GetNext(&url_string);
+
   [[AtomApplication sharedApplication]
       setCurrentActivity:base::SysUTF8ToNSString(type)
-            withUserInfo:DictionaryValueToNSDictionary(user_info)];
+            withUserInfo:DictionaryValueToNSDictionary(user_info)
+          withWebpageURL:net::NSURLWithGURL(GURL(url_string))];
```
Comment (danhp) [inline atom/browser/browser_mac.mm]: NSURL is now `nil` if `url` is invalid (i.e. not specified).
See [NSURLWithGURL](https://code.google.com/p/chromium/codesearch#chromium/src/net/base/mac/url_conversions.h&q=NSURLWithGURL&sq=package:chromium&type=cs&l=19)

File: atom/browser/mac/atom_application.mm [hunk 4]
```diff
@@ -29,10 +29,16 @@ - (void)setHandlingSendEvent:(BOOL)handlingSendEvent {
...
 - (void)setCurrentActivity:(NSString*)type
-              withUserInfo:(NSDictionary*)userInfo {
+              withUserInfo:(NSDictionary*)userInfo
+            withWebpageURL:(NSURL*)webpageURL {
   currentActivity_ = base::scoped_nsobject<NSUserActivity>(
       [[NSUserActivity alloc] initWithActivityType:type]);
   [currentActivity_ setUserInfo:userInfo];
+
+  // Only set the webpageURL if it is valid
+  if (webpageURL && [webpageURL scheme] && [webpageURL host])
```
Comment (zcbenz) [inline atom/browser/mac/atom_application.mm]: We probably don't need the check anymore, `webpageURL` is guaranteed to be either `nil` or a valid URL, and both are safe to be passed to `setWebpageURL`.

**Comment by danhp** [comment]
Done 😄

**Comment by zcbenz** [comment]
👍

**Comment by dasilvacontin** [comment]
Looking forward to playing with this – thanks! ❤️

**Pull request merged**